### PR TITLE
Better publish

### PR DESCRIPTION
**Pull request by arcanis**
This is a fairly large PR that makes it possible to finally publish all the packages in one go to the registry.

- It introduces a new set of commands under the form of a plugin: `yarn version major|minor|patch`. I tried something quite different from the existing approach (maybe I should rename them into something else?).

  High level overview: running `yarn version major` will set a new value in the `version:next` field. Then, when running `yarn version apply`, Yarn will locate all the `version:next` fields and update the matching `version` field, remove the `version:next` fields, and update all the workspace dependencies across the project to match the new version (so `workspace:^1.0.0` would become `workspace:^2.0.0`).

  - It remains to be seen how over-engineered the auto-upgrade mechanism is. The `workspace:` protocol is already replaced at publish-time thanks to @bgotink's work, so this might a case of duplicate feature (especially since it causes a bunch of changes in the workspaces everytime we bump it).

  - The main reason I implemented this was that I wanted to be able to run `yarn version patch` multiple times without bumping the patch number multiple times and without having to keep track of which version number had already been bumped. With this new approach all PRs can run the `version` command on all the packages that will need to be republished, and they will be buffered until I run a release cycle.

  - Related to the previous point, since we keep both the old version and new version, we can be smart enough not to bump the version unless strictly needed (for example if we run `yarn version major` then `yarn version patch`, only the major will be bumped until the changes get persisted via `yarn version apply`).

- I made a few mostly cosmetic changes to the `foreach` command. A few are a bit more complex (cc @deini):

  - I now print something when running with `-v,--verbose` and nothing was printed by the program. I noticed that `tsc` doesn't print anything when it succeeds, and it's hard to keep track of what happens.

  - I've split `--with-dependencies` into `--topological` and `--topological-dev`. This first one only considers `dependencies` when resolving the dependencies, the second considers both `dependencies` and `devDependencies`.

### CODE REVIEW

**Comment by deini** [comment]
`workspaces foreach` changes look good. Not a fan of the new colors but I can live with them:
image

File: packages/plugin-npm-cli/sources/commands/npm/publish.ts [hunk 1]
```diff
@@ -48,8 +48,26 @@ export default (clipanion: any, pluginConfiguration: PluginConfiguration) => cli
 
     // We store it so that TS knows that it's non-null
     const ident = workspace.manifest.name;
+    const version = workspace.manifest.version;
 
     const report = await StreamReport.start({configuration, stdout}, async report => {
+      // Not an error if --tolerate-republish is set
+      if (tolerateRepublish) {
+        const registryData = await npmHttpUtils.get(npmHttpUtils.getIdentUrl(ident), {
```
Comment (bgotink) [inline packages/plugin-npm-cli/sources/commands/npm/publish.ts]: this implies you can't use `--tolerate-republish` for the initial publish of a package, is that by design?

File: packages/plugin-version/sources/commands/version/apply.ts [hunk 2]
```diff
@@ -0,0 +1,147 @@
+import {WorkspaceRequiredError}                                                     from '@berry/cli';
+import {AllDependencies, Cache, Configuration, IdentHash, LightReport, Manifest}    from '@berry/core';
+import {MessageName, PluginConfiguration, Project, StreamReport, WorkspaceResolver} from '@berry/core';
+import {Workspace, structUtils}                                                     from '@berry/core';
+import {PortablePath}                                                               from '@berry/fslib';
+import {Writable}                                                                   from 'stream';
+
+import * as versionUtils                                                            from '../../versionUtils';
+
+const SUPPORTED_UPGRADE_REGEXP = /^([~^]?)[0-9]+\.[0-9]+\.[0-9]+$/;
```
Comment (bgotink) [inline packages/plugin-version/sources/commands/version/apply.ts]: so pre-release versions aren't supported?

**Comment by bgotink** [comment]
Would it make sense to allow custom version numbers as well in the `yarn version` command? Right now there's no way for a developer to publish a prerelease other than editing the version manually.

File: packages/plugin-version/sources/commands/version/apply.ts (hunk 2 above)
Comment (arcanis) [reply packages/plugin-version/sources/commands/version/apply.ts]: Ideally it should be `[~^]{{whatever semver allows}}` 🤔 I'll try to use a more complex regexp.

Ref https://github.com/semver/semver/issues/232

File: packages/plugin-npm-cli/sources/commands/npm/publish.ts (hunk 1 above)
Comment (arcanis) [reply packages/plugin-npm-cli/sources/commands/npm/publish.ts]: Good point, I need to add a guard against 404s 👍

File: packages/plugin-version/sources/commands/version/apply.ts (hunk 2 above)
Comment (arcanis) [reply packages/plugin-version/sources/commands/version/apply.ts]: Ideally it should be `[~^]{{whatever semver allows}}` 🤔 I'll try to use a more complex regexp.

Ref https://github.com/semver/semver/issues/232

**Comment by arcanis** [comment]
> Would it make sense to allow custom version numbers as well in the `yarn version` command? Right now there's no way for a developer to publish a prerelease other than editing the version manually.

👍 I'll add support in a followup PR for:

- `yarn version <semver>` which will set the version to a specific value
- `yarn version` which will ask the user to type a new version number (similar to what happens today)

**Pull request merged**